### PR TITLE
add subset and weights to CategoricalArray arguments

### PR DESCRIPTION
**Pull request by bkamins**
add `subset` and `weights` to `freqtable` for `CategoricalArrays`

### CODE REVIEW

File: src/freqtable.jl [hunk 1]
```diff
@@ -117,14 +128,17 @@ function _freqtable{n}(x::NTuple{n, AbstractCategoricalVector}, skipmissing::Boo
...
+            a[el] += weights[i]
         end
     end
 
     NamedArray(a, lev, ntuple(i -> "Dim$i", n))
 end
 
-freqtable(x::AbstractCategoricalVector...; skipmissing::Bool = false) = _freqtable(x, skipmissing)
+freqtable(x::AbstractCategoricalVector...; skipmissing::Bool = false,
+                   weights::AbstractVector{T} = UnitWeights(),
```
Comment (nalimilan) [inline src/freqtable.jl]: Incorrect alignment.

File: test/freqtable.jl [hunk 2]
```diff
@@ -92,17 +101,30 @@ tabc = freqtable(mcx, mcy, skipmissing=true)
...
+    @test tab == [28 22
+                   3 47
+                   1 49]
+    @test names(tab) == [["setosa", "versicolor", "virginica"], [false, true]]
+    tab = freqtable(iris, :Species, :LongSepal, subset=iris[:PetalLength] .< 4.0)
+    @test tab == [28 22
+                   3  8]
+    @test names(tab) == [["setosa", "versicolor"], [false, true]]
+
+    i
```
Comment (nalimilan) [inline test/freqtable.jl]: ?

Comment (bkamins) [reply test/freqtable.jl]: fixed

File: src/freqtable.jl [hunk 3]
```diff
@@ -104,6 +112,10 @@ function _freqtable{n}(x::NTuple{n, AbstractCategoricalVector}, skipmissing::Boo
 	    end
 	end
 
+    if !isa(weights, UnitWeights) && length(weights) != leb[1]
```
Comment (nalimilan) [inline src/freqtable.jl]: `len` or `lev`? :-)

Comment (bkamins) [reply src/freqtable.jl]: len

Comment (nalimilan) [reply src/freqtable.jl]: Think about the planet, run tests locally before pushing! ;-)

Comment (bkamins) [reply src/freqtable.jl]: sorry - I got into a bad habit as tests often failed; have to readjust.
E.g. I have this problem: https://github.com/JuliaData/DataFrames.jl/pull/1320#issuecomment-353610826 with all the PR I made with DataFrames.

Comment (nalimilan) [reply src/freqtable.jl]: Hmm, better go back to the latest release until we sort this out...

Comment (bkamins) [reply src/freqtable.jl]: I did - the commit I now pushed goes cleanly under DataFrames 0.11.2.

File: src/freqtable.jl [hunk 4]
```diff
@@ -82,15 +82,23 @@ function _freqtable{T<:Real}(x::Tuple,
...
 
 # Internal function needed for now so that n is inferred
-function _freqtable{n}(x::NTuple{n, AbstractCategoricalVector}, skipmissing::Bool = false)
+function _freqtable(x::NTuple{n, AbstractCategoricalVector}, skipmissing::Bool = false,
+                    weights::AbstractVector{<:Real} = UnitWeights(),
+                    subset::Union{Void, AbstractVector{Int}, AbstractVector{Bool}} = nothing) where n
     n == 0 && throw(ArgumentError("at least one argument must be provided"))
+
+    if !isa(subset, Void)
+        x = map(y -> y[subset], x)
```
Comment (nalimilan) [inline src/freqtable.jl]: BTW, that's not related to this PR, but it would make sense to use `view` here and in the other function instead (that would imply using a different name than `x` to avoid a type-instability).

**Comment by nalimilan** [comment]
Thanks!

File: src/freqtable.jl (hunk 4 above)
Comment (bkamins) [reply src/freqtable.jl]: well - `view` can have mixed consequences. I would leave it as is.

Comment (nalimilan) [reply src/freqtable.jl]: In terms of performance? In the present case, since we only iterate over the arrays once, taking a view should always be faster I think.

Comment (bkamins) [reply src/freqtable.jl]: As a side note check out e.g. the performance of:
```
using BenchmarkTools

function f1(x,t)
    y = view(x,t)
    z = 0.0
    for v in y
        z += v
    end
end

function f2(x,t)
    y = x[t]
        z = 0.0
    for v in y
        z += v
    end
end

x = rand(10^6)
t = trues(10^6)

@benchmark f1($x,$t)
@benchmark f2($x,$t)
```

It is an extreme example of course.
On Julia 0.6 `f2` is 2x faser, on Julia 0.7 it is 5% faster, so negligible.

If we take small subsets using booleans (e.g. 10 elements out of 10^6) aslo `f2` is faster, but only marginally.

When range index like `1:10^5` is used `view is faster, but it is significant only if we select a large portion of the original vector.

In summary - `view` is better most of the time but sometimes it is not. We can change it to `view` then 😄 .

Comment (nalimilan) [reply src/freqtable.jl]: Good to have numbers to back the discussion. So indeed it sounds worth using views given that Julia 0.7 will soon be out. An unmeasured advantage is that you need less memory, which isn't negligible if you cross several very large vectors.
